text/0014-is-authorized-batch.md

```diff
@@ -0,0 +1,247 @@
+
+fn list_photos(principal: EntityUid, context: Context) -> Vec<Photo> {
+    // not shown: obtaining list of photo IDs
+    let photo_ids: Vec<EntityUid> = ...;
+    let requests = vec![];
+
+    for id in photo_ids {
+        requests.push(Request::new(Some(principal), Some(ACTION_LIST), Some(id), context))
+    }
+
+    // not shown: obtaining policy set and entities
+
+    let authorizer = Authorizer::new();
+
+    let responses = authorizer.is_authorized(&requests, policy_set, entities);
+
+    let photos = vec![];
+
+    // we should have a request for each photo ID
+    // (in order) and a response for each request,
+    // so transitively we have a response for each
+    // photo ID.
+    for (photo_id, response) in photo_ids.iter().zip(responses.iter_mut()) {
+        match response.decision() {
+            Decision::Allow => {
+                let photo = ...;
+                photos.push(photo)
+            },
+            Decision::Deny => {
+                // skip
+            }
+        }
+    };
+
+    photos
+}
+```
+
+## Motivation
+
+At time of writing at least three language bindings exist
+([ref](https://github.com/cedar-policy/cedar-awesome#language-and-platform-integrations))
+that provide mechanisms to call the Cedar Authorizer's
+[`is_authorized` method](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L300).
+
+While the majority of requests likely operate on a single resource and can be
+handled efficiently by the current implementation, some requests (especially
+`list` requests) may need to check authorization against multiple resources, and
+it is inefficient to do a full round-trip for every request. Similarly, a
+higher-level operation may involve multiple lower-level actions on multiple
+resources.
+
+In naive comparison testing with the
+[k9securityio/cedar-py](https://github.com/k9securityio/cedar-py)
+`is_authorized` implementation against a
+[very naive `is_authorized_batch` implementation](https://github.com/glb/cedar-py/commit/98bf3e1fa2e90bb6f7326cbd846a3293fdbd960b),
+I observed speedups of at least 5x for even simple cases, as documented in
+[k9securityio/cedar-py#13](https://github.com/k9securityio/cedar-py/issues/13).
+
+@skuenzli also
+[reported](https://github.com/k9securityio/cedar-py/issues/13#issuecomment-1629790294)
+that:
+
+> The time observed in Python is consistently 2 orders of magnitude greater than
+> the authorizer.is_authorized call. These particular requests use a simple
+> policy, a simple schema, and ~170 entities.
+
+The initial thought was to simply create an `is_authorized_batch` API for the
+`cedar-py` Python binding, but in discussion in the Cedar Policy Language Slack
+there was support from @aaronjeline and @andrewmwells-amazon for exposing this
+API in Cedar core so that it could be picked up more easily by other language
+bindings, as they expect those bindings to have similar needs and performance
+concerns.
```

Comment:
If you don't mind me expanding your RFC, I can also include some details about the inefficiencies of constructing an `Evaluator` per `is_authorized` call.

Reply:
>Is there a mechanism that a person could use to lazy-load entities as they're needed, or is that just a whole lot of headaches that should be avoided?

Not currently, but maybe something like https://github.com/cedar-policy/rfcs/pull/16 would work?

Reply:
For inefficiencies of constructing an `Evaluator` per `is_authorized` call, see https://github.com/cedar-policy/cedar/issues/227, which is a change we're considering independent of this RFC (and might reduce the motivation for this RFC a little)